<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="description" content="">
    <meta name="author" content="">
    <link rel="shortcut icon" href="../../assets/ico/favicon.png">

    <title>Project proposal</title>

    <!-- Bootstrap core CSS -->
    <link href="bootstrap/css/bootstrap.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="starter-template.css" rel="stylesheet">

    <!-- HTML5 shim and Respond.js IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="../../assets/js/html5shiv.js"></script>
      <script src="../../assets/js/respond.min.js"></script>
    <![endif]-->
  </head>

  <body>

    <div class="navbar navbar-inverse navbar-fixed-top">
      <div class="container">
        <div class="navbar-header">
          <button type="button" class="navbar-toggle" data-toggle="collapse" data-target=".navbar-collapse">
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
            <span class="icon-bar"></span>
          </button>
          <a class="navbar-brand" href="#">Project report</a>
        </div>
        <div class="collapse navbar-collapse">
          <ul class="nav navbar-nav">
            <li class="active"><a href="#title">Title</a></li>
            <li><a href="#team-members">Team members</a></li>
            <li><a href="#abstract">Abstract</a></li>
            <li><a href="#results">Results</a></li>
            <li><a href="#methods">Methods</a></li>
            <li><a href="#discussion">Discussion</a></li>
          </ul>
        </div><!--/.nav-collapse -->
      </div>
    </div>

    <div class="container">

      <div class="starter-template">
        <h1 id="title">Title</h1>
        <p>Pairwise sequence alignment visualization</p>
        <h1 id="team-members">Team members</h1>
        <ul class="list-group">
          <li class="list-group-item">
            Andrew McKinlay
          </li>
        </ul>
        <h1 id="abstract">Abstract</h1>
        <p>Students learning bioinformatics for the first time often find difficulty rationalizing algorithms. Dynamic programming (DP) algorithms, for example, exhibit barriers for students trying to understand sequence comparison. This project aims to ease the transition into bioinformatics for students with a computer science and biology background alike. A web-based demonstration of alignment algorithms through visualization will present an aid to students endeavoring into such topics. The project implements a visualization for local and global alignment algorithms, implementing Needleman-Wunsch and Smith-Waterman, in particular.</p>
        <h1 id="results">Results</h1>
        <p>The visualization takes the form a simple client-side web app. Users will need to have JavaScript enabled. The app consists of two parts, the code to solve dynamic programming pairwise alignments and code to display information generated by the algorithm and accept inputs. When a user inputs two sequences, the alignment and scoring information is displayed below. The matrix denotes scores graphically, each cell in the matrix receiving a proportionally-sized shape to denote its score. As an alignment score increases at any point in the matrix, the corresponding circle diameter increases. Negative scores in the global alignment are reported in a different color than positive scores. The user can then interact with the matrix by hovering over arbitrary points to show a backtracking path. This path traces the matches/mismatches/indels that occurred up to that point in the optimal alignment. Obviously, when a global alignment is performed, this path continues to the origin of the matrix, while in a local alignment, it continues until the point at which the local alignment began. When the user hovers over a score in the matrix, a popover displays the numeric cumulative score at that location as well as whether that location was a match, mismatch, or indel, and the incremental score given. A separate alignment visualization shows the two sequences side-by-side, with indels inserted appropriately. This visualization also responds as the user hovers over different parts of the matrix.</p>

        <p>This whole process occurs in real-time. Whenever the user edits any part of the sequence, the output is shown immediately. Whenever the user hovers over the matrix, backtracing and two-dimensional alignments are shown immediately.</p>
        <h1 id="methods">Methods</h1>
        <p>JavaScript is not known to be the fastest programming language, although performance has increased dramatically with improvements in virtual machine technology. Performance manifests an obstacle when trying to implement a real-time application involving an inherently slow algorithm, such as O(n^2) dynamic programming alignment.</p>

        <p>The underlying matrix representation used in this project, however, does not necessarily run into language impediments. The representation consists of a nested two-dimensional array of objects. Using nested arrays and objects is not a cache friendly approach, requiring at least one dereference to noncontinuous memory on each element access. However, maintaining objects is convenient for backtracking. Each score-containing matrix element contains a reference to a previous element. More efficient approaches are addressed in the discussion. To enhance interactivity, the entire DP matrix is preserved, requiring quadratic space.</p>

        <p>The underlying algorithm allows incremental solving. To increase efficiency, when the user appends or changes information at the tail of either of the sequences (or both), the algorithm avoids rescoring the entire matrix. Only the tail ends of the matrix need to be rescored. The underlying scoring matrix used to create the alignment can also be modified in real-time. The user can edit the score given to a matrch/mismatch/indel and the alignment will update.</p>

        <p>Whenever the length of any of the sequences is reduced, the matrix must shrink. Careful consideration must be taken to manually remove parts of the underlying array representation. JS arrays can be assigned at any arbitrary index, so it is trivial to add elements, but we must explicitly splice the array to remove elements.</p>

        <p>The display of the alignment matrix to the user updates as the underlying representation updates. The display consists of scalable vector graphics (SVG) representing the alignment and the matrix. D3.js is used to bind representational data to the display. Styles and attributes of the XML elements composing the SVGs take functions as arguments. When an array of data is bound to a selection, each individual datum is then passed to a corresponding function and the style or attributes are computed using this datum. A selection consists of a set of DOM elements identified by a CSS selector. When data is bound to the selection, the cardinality of the set of DOM elements may be less than, greater than, or equal to the length of the data. In the first case, d3.js provides an additional method on the selection “enter(),” which returns an imaginary selection of DOM elements that can be styled or attributed as appropriate before being inserted into the document. In the last case, there is a method “exit(),” which returns a selection of DOM elements that no longer have data bound to them. These elements can then be easily removed from the view. The middle case is the “update” selection. There is no need to insert or remove DOM elements in this case, however, existing elements will be styled and attributed according to the new data and functions previously defined. This concept of enter, exit, and update selections are unique to d3.js. It is referred to as a data “join.” Joins are ephemeral and data is only bound for the duration of the selection. No data is actually bound to the DOM elements themselves.</p>

        <p>For the visualization of the alignment matrix, for example, the “size” attribute of an SVG “<circle>” element is set via a function that takes the score and returns the size of proportional circle to display. In this way, numerical scores can be translated into graphical areas. D3.js's built-in scaling utilities ensure proper scaling no matter the maximum score for the alignment. Another example occurs when the user appends to a sequence. The new row or column added to the underlying matrix representation must be forwarded to the matrix view. In this case, a nested data join links the two-dimensional array with a new set of “<circle>” elements in the “enter” selection. Another case occurs when the user modifies the scoring matrix, in this case the new scores are joined and the “enter” and “exit” selections are empty, so no new DOM elements are appended.</p>

        <p>When the user modifies any of the input sequences, the alignment matrix and visualizations update on keyup. This helps to give the feeling of real-time, since the user does not need to press any buttons for the the visualization to update. However, this can cause unresponsiveness for larger sequences.</p>
        <h1 id="discussion">Discussion</h1>
        <p>Unresponsiveness occurs when the browser's JS event loop is busy. For large alignments, the DP algorithm's computational unfriendliness negatively affects user interaction. The code for this project is all synchronous. The instability could be alleviated by using web workers (the proposed standard for parallelism in the browser). Whenever the user modified a sequence a web worker would spawn, doing the DP task asynchronously to not interrupt the user interface. Obviously, additional considerations would have to be taken into account in order to manage the spawned workers. For example, if the user made two modifications in a row, we would want to terminate the first worker.</p>

        <p>At the moment, this project is only a proof of concept. It only supports DNA sequences, although it could easily be extended to support any character sequences given a supporting scoring matrix. The scoring matrix needs to be more editable, as currently its dimensions are fixed. Work also needs to be done to provide default scoring matrices for polypeptide alignment.</p>

        <p>The unresponsiveness that occurs with large alignments in the user interface is an indicator of performance in general. Work needs to be done to profile the code. The underlying DP algorithm implementation may not necessarily be the bottleneck. Data joins when composing the view involve computing attributes and appending/removing DOM elements. Since these operations occur on every update, special examination needs to be performed to identify performance deficiencies.</p>

        <p>The underlying DP algorithm implementation itself poses challenges. Due to JavaScript's uniqueness and the optimizations made by its implementation, knowing data representation at run time can be difficult. Certainly, as mentioned in the methods section, using objects to represent each element in the alignment matrix is a bad idea for both space efficiency and locality. Traversing a nested array when computing alignment scores should be as cache-friendly as possible. As we need to maintain backtracking information at every cell, an alternative matrix could be used for maintaining this data. Furthermore, a more efficient representation of nested arrays should be utilized, especially if the visualization is extended to multiple sequences alignments. Ndarrays is an efficient implementation of n-dimensional arrays. The data remains in a flat, one-dimensional, array. Nesting is available through accessors which provide multidimensional views to the underlying data. This alternative approach leads to much better locality for certain array structures and should be integrated into the project.</p>

        <p>More interactivity should be built into the project. In addition to being able to interact with the alignment matrix, the user should be able to interact with the visualization of the alignment/sequences itself/themselves. A slider over one of the sequences could act as a view-port for a local alignment. As the user drags the slider, a local alignment with the other sequence could be produced in real-time (although, in order for this to occur quickly enough, the alignments would have to be computed ahead of time). Currently, gaps are inserted in the alignment whenever indels occur. We could avoid gaps altogether to preserve the natural representation of the sequence and instead draw paths between monomers (similar to ciros), except in real time. Instead of a local alignment, we could use the view-port to match inversions. As the user slides the re-sizable view-port, inversions are highlighted in the secondary sequence.</p>

        <p>Once performance is improved in the project, it will become much more educative. It can be difficult to demonstrate patterns in short sequences. To accommodate any improvements, the views (visualizations) must also become more malleable. Currently, each cell in the scoring matrix is assigned a fixed-size. Obviously, when the matrix size increases, the matrix does not re-size to fit the screen. This could be fixed in the same way that circles are scaled to the max score. The matrix dimensions should be scaled to the max size of the DOM container. The same goes for the side-by-side visualization of the sequences/alignment. Instead of using circles of varying size to represent score, perhaps color or transparency should be used. This would scale much better to sequences of various size and fill the entire cell, leaving less whitespace and improving the appeal of the visualization.</p>

        <p>Further work needs to go into developing dot-matrix and word method alignments, since most protein alignment algorithms depend on the later. Additional metrics should be developed in addition to the boring two-dimensional dynamic programming matrix. Statistics about in-browser alignment performance should be graphed. Step-by-step instructions for various steps in the alignment process should be explicitly displayed – along with relevant equations. There are alternative variations upon the vanilla scoring matrix, including bounded alignments and algorithms that forgo O(n^2) space requirements at the expense of backtracking. Recursive methods could even be visualized by displaying a recursion tree (with memoized calls colored similarly). Visualization details should also be placed less on the algorithmic structures and more on the alignment itself. There are ways to visualize alignments: visualizations of the underlying algorithm or visualizations of the aligned sequences, themselves. Although it is very difficult to figure out novel ways of displaying an alignment among sequences with intuitive semantics, real-time interactivity may alleviate these frustrations.</p>
      </div>

    </div><!-- /.container -->


    <!-- Bootstrap core JavaScript
    ================================================== -->
    <!-- Placed at the end of the document so the pages load faster -->
    <script src="jquery.min.js"></script>
    <script src="bootstrap/js/bootstrap.min.js"></script>
  </body>
</html>
